Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,256 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.CategoryTheory.Adjunction.Limits
+import Mathlib.LinearAlgebra.Finsupp
+
+/-!
+# Tensor-Hom adjunction
+Consider two commutative rings `R` and `S` and `X` an `(S, R)`-bimodule.
+Consider the tensor functor `(X ⊗[R] .)` from the category of left `R`-module to the category of
+left `S`-module and the hom functor `(X →ₗ[S] ·)` from the category of left `S`-module to the
+category of left `R`-module. They form an adjunction. In particular we have that
+```
+Hom_S(X⊗[R]Y, Z) ≃ Hom_R(Y, Hom_S(X, Z))
+```
+
+## Implementation notes
+
+1. Order of arguments
+```
+Hom_S(Y⊗[R]X, Z) ≃ Hom_R(Y, Hom_S(X, Z))
+```
+is perhaps more natural to work with because all the argument have the same order.
+But currently mathlib4 does not know `Y⊗[R] X` is an right `S`-module as well.
+
+2. Why not use the [`Tower` file](Mathlib/LinearAlgebra/TensorProduct/Tower.lean`)
+
+In our setting `X` is an `(R, S)`-bimodule and `Y` an `R`-module and `Z` an `S`-module
+so to use the `Tower` file, we need `S` to be an `R`-algebra which is a luxury we do not have.
+But note that in `Tower` file, `curry` and `uncurry` are both tri-linear maps. So `Tower` file
+allows interplay of 3 rings which is not allowed in this file.
+
+-/
+
+universe u u' v v' v''
+
+open TensorProduct CategoryTheory
+
+variable (R : Type u) (S : Type u') (X : Type v)
+variable [CommRing R] [Ring S]
+variable [AddCommGroup X] [Module Rᵐᵒᵖ X] [Module S X] [SMulCommClass S Rᵐᵒᵖ X]
+
+open MulOpposite
+
+instance leftMod_eq_rightMod : Module R X :=
+  Module.compHom X ((RingHom.id R).toOpposite fun _ _ => mul_comm _ _)
```

Comment:
This instance from `[CommRing R] [Module Rᵐᵒᵖ X]` to `Module R X` isn't safe as it forms a diamond with [Semiring.toOppositeModule](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Basic.html#Semiring.toOppositeModule).
Lean 4 seems to have trouble with `local instance` so maybe we need to use `local attribute [instance]`.
Of course, when we generalize tensor products it would take `Module Rᵐᵒᵖ X` directly .

Reply:
I think it may be nice to ensure that any `Module Rᵐᵒᵖ X` instance we register for a CommRing R is always "defeq" to the `Module R X` instance; e.g. `Module Rᵐᵒᵖ R` would also be left multiplication rather than right multiplication, which (I think) could be achieved with a higher priority instance for CommRings that override the general Semiring.toOppositeModule. We could even have an instance from `[CommRing R] [Module R X]` to `Module Rᵐᵒᵖ X`. This would also render issue #7152 void, I think.

Reply:
Instance priority is never suitable as a means of ensuring coherent instance resolution; any lemmas about the low-priority instance can introduce the "wrong" one.

I don't understand the motivation for this instance anyway; if you really think you need it, you should instead be writing `[Module Rᵐᵒᵖ X] [Module R X] [IsCentralScalar R X]`.

Reply:
> Instance priority is never suitable as a means of ensuring coherent instance resolution; any lemmas about the "wrong" low-priority instance can introduce it back into the goal.

Okay!

> I don't understand the motivation for this instance anyway; if you really think you need it, you should instead be writing `[Module Rᵐᵒᵖ X] [Module R X] [IsCentralScalar R X]`.

The idea is that `[Module Rᵐᵒᵖ X]` is the morally correct assumption, but mathlib's TensorProduct uses an `R`-action rather than `Rᵐᵒᵖ`-action on the first factor, so we currently need `leftMod_eq_rightMod` for some existing instances to apply. Once we refactor TensorProduct` and switch to using `Rᵐᵒᵖ`-action in the existing instances, `leftMod_eq_rightMod` will no longer be needed. I think adding a local instance is better than adding assumptions, as it doesn't change the interface.

Reply:
I think we should write the interface for the thing we actually have, not add hacks that make it seem like we have the one that we wish we had. It's not mathematically an extra assumption anyway, since we have `CommRing R`.

Reply:
If we want to fit the current interface, we would use `[Module R X]` as the sole assumption and derive `Module Rᵐᵒᵖ X` from it and we'd just be reversing the direction of this instance rather than removing it. Alternatively, the current mathlib practice would be doing this `mop` conversion under the hood (since we always have a CommSemiring when talking about tensor products), but if we do it here we'd just make our future refactoring job harder, I think.

Reply:
I might start experimenting with refactoring TensorProduct tonight and open a draft PR. @jjaassoonn's #8536 seems to be a working towards that but doesn't change the definition of TensorProduct yet; maybe he has some further progress?

Reply:
Forgot to say: if you use `[Module Rᵐᵒᵖ X] [Module R X] [IsCentralScalar R X]`, every caller of `TensorProduct.hcurry` would have to supply those assumptions, probably as local or inlined instances.